### ISSUE TITLE
Interface to use BOUT++ indices with PETSc objects

### ISSUE THREAD

**Issue by cmacmackin** [boutproject/BOUT-dev#1771]
I've been tasked by @johnomotani with implementing a 3D AMG Laplace solver using PETSc. While studying the existing PETSc Laplace implementations I was struck by the somewhat complicated and error-prone approaches to setting elements of the matrix operator and, to a lesser extent, PETSc vectors. I also noticed that these implementations do not make use of the current [recommended way to iterate over fields](https://bout-dev.readthedocs.io/en/latest/developer_docs/data_types.html#iterating-over-fields). I think there might be a better way to do it for my work but, before I started implementing it, I wanted to make sure something like this doesn't already exist. I also thought it might be useful to gauge interest in such a feature to determine how general I should make it and whether it should be publicly accessible beyond the 3D AMG implementation I will be working on.

The ideas is to create `PetscMatrix` and `PetscVector` classes. These would look something like as follows:

```C++
class PetscMatrix {
public:
  // Take BOUT++ indices and return an object to access the corresponding element in 
  // a PETSc matrix.
  PetscElement operator[](int rowIndex, int colIndex);
  void clear();
  void assemble();
private:
  Mat *matrix;
  // Various other information needed to handle indexing
}

class PetscVector {
 // Take BOUT++ indix and return an object to access the corresponding element in 
  // a PETSc vector.
  PetscElement operator[](int index);
  void clear();
  void assemble();
private:
  Vec *vector;
  // Various other information needed to handle indexing
}

class PetscElement {
public:
  // Handle setting this element of the matrix/vector
  virtual BoutReal operator=(BoutReal val);
  virtual BoutReal operator+=(BoutReal val);
  // Get the value of this element of the matrix/vector for use in BOUT++
  virtual BoutReal getValue();
}

// Plus subclasses of PetscElement for matrix and vector elements...

// Some sort of krylovSolve() function might also be useful. For example:
PetscVector krylovSolve(PetscMatrix A, PetscVector b, PetscVector x0);
// Plus whatever other settings need to be passed in for preconditioning, etc.
```

This should make it much easier to set up differentiation matrices. For example, computing a second-order first derivative in _x_ would look something like this.
```C++
// We want to solve for b in Ab = c
PetscMatrix A(); // Pass in information on memory allocation in a sparse matrix and on indexing
PetscVector b(), c(); // Pass in information on indexing (e.g., a pointer to the global mesh)

Field3D b_field(0.0), c_field(1.0);
// Boundaries will be handled in separate loops
BOUT_FOR(i, a_field.getMesh()->getRegion3D("RGN_NOBNDRY")) {
  A[i, i.xp] = 0.5;
  A[i, i.xm] = -0.5;
  b[i] = b_field[i];
  c[i] = c_field[i];
}
A.assemble();
b.assemble();
c.assemble();
// Some sort of call to use the PETSc solvers...
```

**Comment by d7919** [comment]
@cmacmackin I agree the current approach is not that clear and could be quite error prone and had thought of a similar approach to that which you describe but never got around to doing anything about it. We also have a little bit of similar code in the imex-bdf solver (integrator) and possible the snes one as well, so ensuring the approach is compatible with the uses there would be helpful.

(thanks for Ford as well!)

**Comment by cmacmackin** [comment]
Glad to hear this approach would be useful to others. Where in the source tree would it be best to implement these classes? 

Just to clarify, I am not proposing to implement all of the features myself which these other use-cases would need, as that is beyond the scope of the project I've been asked to complete. However, I'm happy to discuss the approach I use to ensure that it does not prevent this being done in future. I hope that's alright.

**Comment by d7919** [comment]
Yes, as long as the design/approach can be extended in the future that should work, @ZedThree and @bendudson are probably best placed to discuss structure etc.

**Comment by bendudson** [comment]
Hi @cmacmackin yes this would be great! I'm afraid I perpetrated the mess which is the current PETSc solvers, and a tidier way to do it would be nice. Your sketch of an interface looks very good. 
Much of the complexity with the current method comes from:
1. The need to preallocate the matrix row sizes (is this no longer needed, or perhaps allocating and filling could be done as part of `assemble`?). This duplicates much of the logic around boundary conditions for allocation, and then filling in the values.
2. Boundary conditions, which can come in several varieties.
3. There is a modest amount of awfulness in `globalIndex` which takes care of the non-trivial way the processors can be connected together. This could be at least wrapped up in this `PetscMatrix` class.

It might be worth a bit of a hunt around for existing C++ interfaces to PETSc. I didn't see anything in a quick search, but perhaps some this has been done already. The python PETSc interface may also be worth looking at.

Perhaps this should go under `src/utils` and `include/bout/utils` ?

**Comment by ZedThree** [comment]
Implemented in #1805
